### PR TITLE
PW-755 Move OpenInvoice to Checkout 

### PR DESCRIPTION
**Pull request by cyattilakiss**
**Description**
Move open invoice to checkout
Put HPP redirect and Validate3d redirect (general handler) together
Put HPP into the authorize flow instead of the order flow
Show gender, date of birth and phone number for open invoice payment
methods all the time - remove the admin settings
Fix house number in the AddressBuilder
Add first name, last name and country code in the CustomerDataBuilder

### CODE REVIEW

File: Controller/Process/Redirect.php [hunk 1]
```diff
@@ -46,71 +46,202 @@ class Redirect extends \Magento\Framework\App\Action\Action
...
+					__('Customer was redirected to bank for 3D-secure validation. Once the shopper authenticated, the order status will be updated accordingly. 
+                        <br />Make sure that your notifications are being processed! 
+                        <br />If the order is stuck on this status, the shopper abandoned the session. The payment can be seen as unsuccessful. 
+                        <br />The order can be automatically cancelled based on the OFFER_CLOSED notification. Please contact Adyen Support to enable this.'))->save();
+				$this->_view->loadLayout();
+				$this->_view->getLayout()->initMessages();
+				$this->_view->renderLayout();
+			}
+
+		} else if ($checkoutAPM) {
```
Comment (Aleffio) [inline Controller/Process/Redirect.php]:
```suggestion
		} else if (!empty($checkoutAPM) && $checkoutAPM) {
```

File: Gateway/Request/AddressDataBuilder.php [hunk 2]
```diff
@@ -63,16 +63,23 @@ public function build(array $buildSubject)
...
+                "houseNumberOrName" => 'N/A',
+                "stateOrProvince" => 'N/A',
                 "country" => "ZZ"
             ];
 
-            if ($billingAddress->getStreetLine1()) {
-                $requestBilling["street"] = $billingAddress->getStreetLine1();
-            }
+			if ($billingAddress->getStreetLine1()) {
+				$address = $this->adyenHelper->getStreet($billingAddress->getStreetLine1());
```
Comment (Aleffio) [inline Gateway/Request/AddressDataBuilder.php]: adyenHelper->getStreet() wants an address as a parameter, it does not work with string

File: Gateway/Request/AddressDataBuilder.php [hunk 3]
```diff
@@ -94,17 +101,35 @@ public function build(array $buildSubject)
...
-            $requestDelivery = ["street" => $shippingAddress->getStreetLine1(),
-                "postalCode" => $shippingAddress->getPostcode(),
-                "city" => $shippingAddress->getCity(),
-                "houseNumberOrName" => '',
-                "stateOrProvince" => $shippingAddress->getRegionCode(),
-                "country" => $shippingAddress->getCountryId()
-            ];
 
+			if ($shippingAddress->getStreetLine1()) {
+				$address = $this->adyenHelper->getStreet($shippingAddress->getStreetLine1());
```
Comment (Aleffio) [inline Gateway/Request/AddressDataBuilder.php]: adyenHelper->getStreet() wants an address as a parameter, it does not work with string

File: Gateway/Request/AddressDataBuilder.php [hunk 4]
```diff
@@ -63,16 +63,23 @@ public function build(array $buildSubject)
         if ($billingAddress) {
 
             $requestBilling = ["street" => "N/A",
-                "postalCode" => '',
+                "postalCode" => 'N/A',
                 "city" => "N/A",
-                "houseNumberOrName" => '',
-                "stateOrProvince" => '',
+                "houseNumberOrName" => 'N/A',
+                "stateOrProvince" => 'N/A',
```
[the remark this answers is not in the document]
Comment (cyattilakiss) [reply Gateway/Request/AddressDataBuilder.php]: If it's empty now we are not redirected to the hpp where we could fill a form with the required fields but get a response in which the API asks for the required info. 

Actually I got the idea from this: https://github.com/Adyen/adyen-magento2/blob/develop/Block/Redirect/Redirect.php#L349

File: view/frontend/web/template/payment/hpp-form.html [hunk 5]
```diff
@@ -69,72 +69,68 @@
...
+                                                title: $t('Telephone'),
+                                                'data-container': getCode() + '-telephone_' + value,
+                                                'data-validate': JSON.stringify({'required-number':true})
+                                            },
+                                            value: telephone"
+                                       data-validate="{required:true}"
+                                />
                             </div>
-                        <!--/ko-->
+                        </div>
```
[the remark this answers is not in the document]
Comment (cyattilakiss) [reply view/frontend/web/template/payment/hpp-form.html]: Looks like only the indention is wrong

File: Gateway/Request/AddressDataBuilder.php [hunk 6]
```diff
@@ -63,16 +63,23 @@ public function build(array $buildSubject)
...
-            }
+			if ($billingAddress->getStreetLine1()) {
+				$address = $this->adyenHelper->getStreet($billingAddress->getStreetLine1());
+
+				if ($address) {
+					$requestBilling["street"] = $address["street"];
+					$requestBilling["houseNumberOrName"] = $address["house_number"];
+				} else {
+					$requestBilling["street"] = $billingAddress->getStreetLine1();
+				}
```
[the remark this answers is not in the document]
Comment (cyattilakiss) [reply Gateway/Request/AddressDataBuilder.php]: yes it is

File: Gateway/Request/AddressDataBuilder.php [hunk 7]
```diff
@@ -94,17 +101,35 @@ public function build(array $buildSubject)
...
 
+			if ($shippingAddress->getStreetLine1()) {
+				$address = $this->adyenHelper->getStreet($shippingAddress->getStreetLine1());
+
+				if ($address) {
+					$requestDelivery["street"] = $address["street"];
+					$requestDelivery["houseNumberOrName"] = $address["house_number"];
+				} else {
+					$requestDelivery["street"] = $shippingAddress->getStreetLine1();
+				}
```
[the remark this answers is not in the document]
Comment (cyattilakiss) [reply Gateway/Request/AddressDataBuilder.php]: yes it is

**Pull request merged**